Adding sass task to wsk

gauntface opened the pull request:
@surma 

Thoughts on directory names, general approach with config and vague structure.

surma commented on package.json:
```diff
@@ -2,7 +2,12 @@
-    "gulp": "gulpjs/gulp#4.0"
+    "gulp": "gulpjs/gulp#4.0",
+    "gulp-autoprefixer": "^3.1.0",
+    "gulp-minify-css": "^1.2.3",
```
`npm WARN deprecated gulp-minify-css@1.2.3: Please use gulp-cssnano instead` LOLs. WDYT?

surma commented on src/gulp-tasks/sass.js:
```diff
@@ -0,0 +1,61 @@
+import sourcemaps from 'gulp-sourcemaps';
+
+const AUTOPREFIXER_BROWSERS = [
+  'ie >= 10',
```
:skull: IE11 plz

surma commented on src/gulp-tasks/sass.js:
```diff
@@ -0,0 +1,61 @@
+  'ios >= 7',
+  'android >= 4.4',
+  'bb >= 10'
+];
```
We could probably bump a few of these in general

surma commented on src/gulp-tasks/sass.js:
```diff
@@ -0,0 +1,61 @@
+  'bb >= 10'
+];
+
+gulp.task('sass', () => {
```
What about the idea I described briefly:

Use ES2015 modules to contain the functions and then hook up these functions in gulp tasks (i.e. the proper gulp4 way) instead of using string identifiers, because that makes these tasks callable from the CLI.

surma commented on src/gulpfile.babel.js:
```diff
@@ -19,20 +19,16 @@
-  cb();
-});
+// Load tasks from the `gulp-tasks` directory
+requireDir('gulp-tasks');
```
This isn‘t very ES2015, is it?

surma commented on src/gulp-tasks/sass.js:
```diff
@@ -0,0 +1,61 @@
+
+  // Only create sourcemaps for non prod
+  if (GLOBAL.config.env !== 'prod') {
+    sassStream.pipe(sourcemaps.init());
```
Not sure, but I *think* this needs to be
```JS
sassStream = sassStream.pipe(sourcemaps.init());
```

Same for the rest,

gauntface commented:
@surma thanks for all the feedback, I think I got most of it.

I changed autoprefixer to use the last 2 versions for chrome, firefox, edge, safari and opera. Not certain if the syntax is correct or not, but the tool runs, will need to investigate a bit further.

Totally right regarding piping the streams.

Also moved to simple functions. This means we'll have to come up with a different way to get watch tasks, but I don't think that'll be too hard to fix.

How did you find that the original minifier was deprecated? Would like to get that tracked somewhere automatically.

Another review would be awesome.

surma commented:
>
> Also moved to simple functions. This means we'll have to come up with a
> different way to get watch tasks, but I don't think that'll be too hard to
> fix.
>
I was thinking:

export default {
  task: function(){},
  watch: function(){},
  clean: function(){},
}

WDYT?

How did you find that the original minifier was deprecated? Would like to
> get that tracked somewhere automatically.
>
npm install got in my face about it. (Has been with MDL for a while now)


> Another review would be awesome.
>
Will do

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/web-starter-kit/pull/813#issuecomment-171979870>
> .
>

gauntface commented:
The only downside of the proposal would be that you would have to use it the other side as:

import sass from './gulp-tasks/sass';

gulp.series([sass.task])

Not a biggy, but would have loved to just had sass, but I honestly can't think of any other option.

Alternative is to do something like:

    import {watch as sassWatch} from './gulp-tasks/sass';
    if (sassWatch) {
      console.log('Sass has a watch task');
    }

If we have a gulp 'watch' task that hunts out all the inidividual watch tasks and runs them, this check isn't too bad, only thing is that we have to get all the files that are available to import and check if a watch function exists.

surma commented:
That’s even better :+1:  Let’s use the ES6 module features available to us.

surma commented on src/gulp-tasks/sass.js:
```diff
@@ -0,0 +1,52 @@
+  // We only want to minify for production builds
+  if (GLOBAL.config.env === 'prod') {
+    sassStream = sassStream.pipe(cssnano());
+  }
```
Stylistic/General Q: Wanna try to avoid stuff like `gulp-if`? I would make the pipeline more obvious (syntactically), but I am not a big fan of the plugin.

gauntface replied:
I'm trying to cut down the number of dependencies and I have hit some small but annoying snags with gulp-if, so if you aren't disgusted with this approach then I'd like to keep, but I agree gulp-if does look nicer.

surma replied:
Totally in favor, was just curious if this was a conscious decision.

surma commented on src/gulp-tasks/sass.js:
```diff
@@ -0,0 +1,52 @@
+/**
+ *
+ *  Web Starter Kit
+ *  Copyright 2015 Google Inc. All rights reserved.
```
2016, yo

gauntface commented:
cc @surma and @addyosmani

surma commented:
LGTM :)
